### DISCUSSION TITLE
Load sql file from disk and bind parameters?

### DISCUSSION THREAD

**Discussion by martinklepsch** [pocketbase/pocketbase#4488] [Q&A] [answered]
Hey! I initially thought I could do this with the **View Collection** feature but then noticed that it only supports fully static SQL queries. So I thought, maybe this could be done using the Extension features (which seem really well designed btw!) but I ran into some issues that I wasn't sure how to debug furhter. 

It looks like with code like this the `file` variable will be a bunch of numbers (bytes?) but not the string thats stored in the file. I'm probably missing something here. Or this is a bug? 

```js
  const file = $os.readFile(__hooks + "/query.sql")
  console.log(file)
  const users =  $app.dao().db()
    .newQuery(file)
    .bind({
      "id": id
    })
    .one(result)
```

**Comment by ganigeorgiev** [comment, marked as answer]
The JSVM `$os.readFile` is a direct mapping of the Go [`os.ReadFile`](https://pkg.go.dev/os#ReadFile).

To get a plain string from the bytes slice you can try converting it with `String.fromCharCode`/`String.fromCodePoint`:

```js
const fileStr = String.fromCharCode(...file)

// with v0.23.0+ this could be replaced with the toString() helper
// (this should also handle properly multi-byte characters, see the comment below)
const fileStr = toString(file)
```

_Edit: note that the `one(result)` doesn't return anything and the query result will be populated into the `result` variable. It throws a regular JS exception on error._

**Reply by martinklepsch** [reply]
Thank you so much for the swift reply! I didn't realize this maps to Go so directly (and am generally a bit unfamiliar with Go) but the solution is easy enough, thank you! 

Have you considered adding parameters to View-type Collections? It seems with the Extension API I can implement this quite easily myself but also seems useful generally perhaps?

**Reply by ganigeorgiev** [reply]
A "View collection" is effectively a plain [SQLite View](https://www.sqlite.org/lang_createview.html) and it cannot be parameterized at the moment without implementing something totally different.

Note that once a View collection is created you can use it to filter/search in it like any other collection:

```js
const records = await pb.collection("viewCollection").getFullList({
    filter: "something = '123'"
})
```

**Reply by martinklepsch** [reply]
Duh! I was kind of forgetting that I can just query the view as it's own table — that completely solves what I needed I believe!

**Reply by martinklepsch** [reply]
This is so nice to work with, thanks a lot! I know you said you don't currently accept donations, but I'd be happy to just send you €100 as a thank you for making Pocketbase and creating such a beautiful API. Send me an email :)

**Comment by martisj** [comment]
Instead of creating a new issue or discussion I thought it made sense to add to this one. How do you make sure the `[]byte` (in GO) or the `number[]` (in JS) is utf-8 encoded? I have a json file with emojis and scandinavian letters in it. Example:
```json
[{
"rating": "good 😊"
}]
```
When running this code:
```js
$app.rootCmd.addCommand(
  new Command({
    use: 'mini-test',
    run: (cmd, args) => {
      const seedDataBuffer = $os.readFile('mini-test-case.json');
      const json = String.fromCharCode(...seedDataBuffer);
      const seedData = JSON.parse(json);

      console.log(JSON.stringify(seedData, null, 2));
    },
  })
);
```
The output is:
```
2024/08/27 01:21:49 [
  {
    "rating": "good ð"
  }
]
```
I would expect the output to be the same as the json file itself. Any suggestions as to how I can make sure to use utf-8 here?

**Reply by ganigeorgiev** [reply]
The resulting string is already utf8 encoded. What you need is probably to convert/combine the individual ut8 single-byte characters into multi-byte unicode characters. In regular JS runtimes you usually would use the [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder) api but this doesn't seem to be supported in goja, so a workaround could be to create an escape sequence from the encoded string and then decode it back, something like:

```js
const json = decodeURIComponent(escape(String.fromCharCode(...seedDataBuffer)))

// should properly print the multi-byte characters
console.log(json)
```

In terms of efficiency I haven't tested it but for small texts it should be fine.

I'll consider with the next release to expose a direct helper (ex. `toString()`) that will convert a byte slice to string as this is supported natively in Go.
_>>> Done. With v0.23.0 the new helper `toString` will superseed the existing `readerToString` and will add support for any type to string conversions (including byte slices)._

**Reply by martisj** [reply]
Phenomenal! Thank you. String encoding/decoding is still a blackbox for me (after 10 years programming). This might be a different issue altogether (let me know if I should file another issue): I can't use the decoded JSON to define a file path to read from with `$filesystem.fileFromPath(pathFromJsonField);` however if I hardcode the path it works. The error I get is `2024/09/03 11:56:28 GoError: stat ./pics/d3f0b7d0-d548-11e8-8ec8-97a4f202983a.jpeg
: no such file or directory` but that file definitely exists.
![image](https://github.com/user-attachments/assets/70865d6b-87bf-4c65-8ea0-6f5d4e06a39a)

Is it because the command file ("./pb_hooks/seedDatabase.pb.js") is in a sub-directory?

**Reply by ganigeorgiev** [reply]
> I can't use the decoded JSON to define a file path to read from with $filesystem.fileFromPath(pathFromJsonField); however if I hardcode the path it works.

I'm not sure I follow but I suspect that your issue is related to the use of relative paths (see the note in https://pocketbase.io/docs/js-overview/#relative-paths).
